File: 백준/Bronze/1952. 달팽이2/달팽이2.py
M, N = map(int, input().split())
cnt = 0
matrix = list([0]*N for _ in range(M))

# 달팽이 이동 델타탐색
dr = [0, 1, 0, -1]
dc = [1, 0, -1, 0]
f = 0
# 시작 좌표
r, c = 0, 0

for i in range(1, M*N+1):
    matrix[r][c] = i
    r += dr[f]
    c += dc[f]
    if r < 0 or r >= M or c < 0 or c >= N or matrix[r][c]:
        r -= dr[f]
        c -= dc[f]
        f = (f+1) % 4
        r += dr[f]
        c += dc[f]
        cnt += 1

print(cnt-1)

File: 백준/Bronze/1952. 달팽이2/달팽이2.js
const readline = require('readline');
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

rl.question('', (input) => {
  const [M, N] = input.split(' ').map(Number);
  let cnt = 0;
  const matrix = Array.from({ length: M }, () => Array(N).fill(0));

  // 달팽이 이동 델타탐색
  const dr = [0, 1, 0, -1];
  const dc = [1, 0, -1, 0];
  let f = 0;
  // 시작 좌표
  let r = 0, c = 0;

  for (let i = 1; i <= M * N; i++) {
    matrix[r][c] = i;
    r += dr[f];
    c += dc[f];
    
    if (r < 0 || r >= M || c < 0 || c >= N || matrix[r][c]) {
      r -= dr[f];
      c -= dc[f];
      f = (f + 1) % 4;
      r += dr[f];
      c += dc[f];
      cnt++;
    }
  }

  console.log(cnt - 1);
  rl.close();
});
